### DISCUSSION TITLE
Best way to embed resources?

### DISCUSSION THREAD

**Discussion by cristianadrielbraun** [AllenDang/giu#399] [Q&A] [answered]
I was thinking about embeding, for example, a few png or svg files to be used as icons in my app.
But, naturally, I don't want to have these files lying around. I would like to enbed them in the binary.

How would you achieve this?

I tried with go:embed, go.rice, tried to get an image from base64, but I was unsuccessful.

**Comment by gucio321** [comment, marked as answer]
what about this?
```golang
package main

import (
	"bytes"
	_ "embed"
	"image"
	"log"

	"github.com/AllenDang/giu"
)

//go:embed gopher.png
var iconData []byte

var icons []image.Image

func loop() {
	giu.SingleWindow().Layout(
		giu.Label("Hello world"),
	)
}

func main() {
	icons = make([]image.Image, 0)

	icon1, _, err := image.Decode(bytes.NewReader(iconData))
	if err != nil {
		log.Fatal(err)
	}

	icons = append(icons, icon1)

	wnd := giu.NewMasterWindow("window icon (#399) [demo]", 640, 480, 0)
	wnd.SetIcon(icons)
	wnd.Run(loop)
}

```

you can embed as much png files, as you want or use embed.FS to embed a whole directory.

**Comment by cristianadrielbraun** [comment]
That was very helpful, thanks!
I was close, but I didn't get the image.Decode part right. Thanks to your example, I was able to do it.

I used a single png, and it wasn't intended as the window icon (I know that function requires a image array)
So, I skipped the array part, and it was enough with this function:

```go
icon1, _, err := image.Decode(bytes.NewReader(iconData))
	if err != nil {
		log.Fatal(err)
	}
```
after that, I can just use icon1 in , for example, a giu.ImageWithRGBA

PS:
I don't know why, but I'm having an "invalid pointer" error with the "ImageButtonWithRGBA".
I think I recall solving this converting the RGBA to Texture, using the  giu.NewTextureFromRgba, and then a giu.ImageButton (which uses a Texture instead of a RGBA), but I couldn't get it working neither now.
Also, it works perfectly with "ImageWithRGBA" (Literally, I just removed the "Button" word from the function). It even has the OnClick function, so is not a big deal for me in my particular situation.
But a button that not behaves like a button (You know, the border, the hover, the click effect, etc) may not be ideal.

Did I just discovered a bug?

**Reply by gucio321** [reply]
yah, it could be a bug. Could you open an issue

**Reply by cristianadrielbraun** [reply]
Of course! I'll do it in a couple of hours
Thanks for the help, the embed solution worked perfectly.

**Comment by cristianadrielbraun** [comment]
This is a small additon, but in case anyone needs this too:
As gucio321 said, if you are embeding a lot of files, you can do it with embed.FS.

in that case, the procedure is similar, but you have to read the file first to a []byte, an then decode.

So, for this file tree:
```
resources/
|-- icon1.png
|-- icon2.png
|-- icon3.png
```
you can:
```golang
//go:embed resources/*
var resourcesFS embed.FS
```

and, to get the icons, you should do something like:
```golang
[...]
func main() {
	icons = make([]image.Image, 0)

	iconAsByte, err := resourcesFS.ReadFile("resources/icon1.png")
	if err != nil {
		log.Fatal(err)
	}
	icon1, _, err := image.Decode(bytes.NewReader(iconAsByte))
	if err != nil {
		log.Fatal(err)
	}
	icons = append(icons, icon1)
[...]
```
Note the extra step using the resourcesFS.ReadFile. That result in the desired []byte variable.

The smart way to load a few would be to write a function to dynamically load the needed file.. maybe something like:

```golang
func setIcon(file string) int {
	iconByte, err := resources.ReadFile(file)
	if err != nil {
		log.Fatal(err)
	}
	iconAsRGBA, _, err := image.Decode(bytes.NewReader(iconByte))
	if err != nil {
		log.Fatal(err)
	}
	icons = append(icons, iconAsRGBA)	
	
	// This return should be safe as long you append to the array from one at the time
	//.. and you should anyways
	return len(icons) - 1 
}
```
That function should be called as:
```golang
icon1Index := setIcon("resources/icon1.png")
icon2Index := setIcon("resources/icon2.png")
icon3Index := setIcon("resources/icon3.png")

```
that will return the index of the appended icon in the slice.
after that you can just acces it along your code as: 

`icons[icon1Index]`

or asign a variable to it:

```golang
var icon1 image.Image
icon1 = icons[icon1Index]
```

PS: That last function ocurred to me as I was writing, so it could be better/safer/smarter ways to do it. If you have one, share it!

I hope this could be useful to someone.

**Reply by gucio321** [reply]
or for better automatication:
```golang
//go:embed // assets/
var iconesDataDir *embed.FS

func loadImages() (result map[string]image.Image) {
  for _, file := range iconesDataDir.ReadDir(".") {
    result[file], _, err := image.Decode(bytes.NewReader(iconesDataDir.ReadFile(file))
  }
  return result
}
```
this should load all the content of assets :grinning:
